### PR TITLE
[compute] [azure_arm] New fixtures, and minor changes to work with new API

### PR DESCRIPTION
**Pull request by SamuelMarks**
### Description
Was careful not to deprecate previous approaches, by using `.get` rather than just replacing the key name to the new one, as not sure if that new one is used in all interfaces from Microsoft.

`datetime` usage also is clearer to read, not to mention still supporting the epoch syntax [if MS still gives that somewhere?].

### Status

- done, ready for review

### Checklist (tick everything that applies)

- [ ] [Code linting](http://libcloud.readthedocs.org/en/latest/development.html#code-style-guide) (required, can be done after the PR checks)
- [x] Documentation
- [x] [Tests](http://libcloud.readthedocs.org/en/latest/testing.html)
- [ ] [ICLA](http://libcloud.readthedocs.org/en/latest/development.html#contributing-bigger-changes) (required for bigger changes)

### CODE REVIEW

**Comment by SamuelMarks** [comment]
@tonybaloney Is there a caching model for libcloud?

The `.list_images` method takes 25+ minutes to populate a pretty-printed 2.1MB JSON file of all the publicly registered Azure images. It's doing 1 request per image. Using cPickle the result was 1.3MB.

If you like, I can put this code into the library as an `ex_cache_dir=` which will set caching to True and put the files in `path.join(ex_cache_dir, 'azure_arm_{location}_images.pkl')`. I'll add a `ex_cache_force_refresh=` to invalidate the cache and reload it.

It takes way too long otherwise :hourglass_flowing_sand:

File: docs/examples/compute/azure_arm/instantiate.py [hunk 1]
```diff
@@ -2,6 +2,6 @@
 from libcloud.compute.providers import get_driver
 
 cls = get_driver(Provider.AZURE_ARM)
-driver = cls(tenant_id='tenant_id',
-             subscription_id='subscription_id',
-             key='application_id', secret='password')
+driver = cls(tenant_id='$tenantId',
```
Comment (tonybaloney) [inline docs/examples/compute/azure_arm/instantiate.py:5]: why are you using dollar signs?

**Comment by tetron** [comment]
I haven't tested these changes but in principle I don't have any objections.

**Review by tonybaloney — CHANGES_REQUESTED** [review]

**Comment by tonybaloney** [comment]
and @SamuelMarks with respect to caching, no libcloud doesn't have a caching mechanism. I had the same issue in testing, it takes a ridiculous amount of time to get full details about images.

**Comment by pquentin** [comment]
@SamuelMarks Can you please separate the " -> ' change in a single commit? And please don't change the license text. We'll need a rebase at some point too.

File: docs/examples/compute/azure_arm/instantiate.py (hunk 1 above)
Comment (SamuelMarks) [reply docs/examples/compute/azure_arm/instantiate.py:5]: Seemed to be a well-known way of saying "hey, this is a variable".

**Comment by SamuelMarks** [comment]
Rebased. The goal is to replace the first 3 CLI commands that require the Azure client installed, with Apache Libcloud code: http://libcloud.readthedocs.io/en/latest/compute/drivers/azure_arm.html

**Comment by SamuelMarks** [comment]
Rebase isn't reflected, making new PR.

**Comment by pquentin** [comment]
Thanks for your work on this. Please tell me when you consider the pull request ready. :)

**Comment by Kami** [comment]
First thank you for your contribution.

This PR is very old and out of date now so I will go ahead and close it.

If it's still relevant, please sync your changes with the latest master / trunk and open a new PR. Thanks!

**Comment by SamuelMarks** [comment]
Merged 581 commits from `apache:trunk` into `SamuelMarks:azure_arm_patches`; so it's now up-to-date.

**Pull request closed without merging**